### PR TITLE
[Model] Gated Graph Neural Network for bAbI tasks

### PR DESCRIPTION
**Pull request by HuXiangkun**
Gated Graph Neural Networks using GatedGraphConv and GlobalAttentionPooling as building blocks. This example contains Node Selection, Graph Classification and Sequence outputs tasks in bAbI dataset.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: examples/pytorch/ggnn/README.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+- Paper link: https://arxiv.org/pdf/1511.05493.pdf
+
+## Dependencies
+- PyTorch 1.0+
+
+- DGL 0.3.1+
+
+## Solving bAbI tasks
+
+- bAbI dataset is generated randomly by [script](https://github.com/facebook/bAbI-tasks). For convenience, we have generated the dataset we used in this example [here](https://s3.us-east-2.amazonaws.com/dgl.ai/models/ggnn_babi_data.zip). Unzip the file to current directory.
```
Comment (mufeili) [inline examples/pytorch/ggnn/README.md]: Maybe you can add code to automatically handle this. See this [example](https://github.com/dmlc/dgl/blob/9c790b114326829044bb6fc8ac73869151210910/python/dgl/data/reddit.py#L12).

File: examples/pytorch/ggnn/README.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+- Paper link: https://arxiv.org/pdf/1511.05493.pdf
+
+## Dependencies
+- PyTorch 1.0+
+
+- DGL 0.3.1+
+
+## Solving bAbI tasks
+
```
Comment (mufeili) [inline examples/pytorch/ggnn/README.md]: Could you please elaborate more about the dataset and task? This does not need to be perfect, but should provide an appropriate context for understanding the scripts and configurations you use below, including:
- What do "ns", "gc", "path_finding" stand for and what will be their semantic meanings in the particular scenario?
- What do settings like `task_id`, `question_id`, `question_type`, `train_num` refer to?

File: examples/pytorch/ggnn/README.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+- bAbI dataset is generated randomly by [script](https://github.com/facebook/bAbI-tasks). For convenience, we have generated the dataset we used in this example [here](https://s3.us-east-2.amazonaws.com/dgl.ai/models/ggnn_babi_data.zip). Unzip the file to current directory.
+
```
Comment (mufeili) [inline examples/pytorch/ggnn/README.md]: The bAbI suite has "20" tasks and you only mention task 4, 15, 16, 18, 19 below. Does it mean that you only supported these 5 tasks? If so, clearly state that. If there are any particular motivations/reasons, include them.

File: examples/pytorch/ggnn/README.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+- For task 4, there are four question types: n, s, w, e, their corresponding question id are ranged from 0 to 3. For each question type, run the following commands.
```
Comment (mufeili) [inline examples/pytorch/ggnn/README.md]: Elaborate a bit about the semantic meanings of the question types. I think this sentence will be nice to quote

> Some tasks have multiple question types, for example Task 4 which has 4 question types: e, s, w, n. For such tasks we simply train a separate GGNN for each task.

*Grammar correction*: "are ranged from" should be "ranges from".

File: examples/pytorch/ggnn/README.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+We can get average accuracy of 100.0 with std 0.0 on 10 different test datasets for both question types.
+
+- For task 19, the *Path Finding* task, there are no question types. Run the following command
+
+```bash
+python train_path_finding.py --train_num=250 --epochs=200
+```
+
+We can get average accuray of 97.8 with std 0.02 on 10 different test datasets.
+
```
Comment (mufeili) [inline examples/pytorch/ggnn/README.md]: Briefly, I got either an  accuracy of "0" or "1" for each task, which is a bit extreme. Include some explanations for reference.

File: examples/pytorch/ggnn/ggsnn.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
+"""
+Gated Graph Sequence Neural Network for sequence outputs
+"""
+
+from dgl.nn.pytorch import GatedGraphConv, GlobalAttentionPooling
+import torch
+from torch import nn
+import dgl
```
Comment (mufeili) [inline examples/pytorch/ggnn/ggsnn.py]: `dgl` unused.

File: examples/pytorch/ggnn/ggnn_gc.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
+"""
+Gated Graph Neural Network module for graph classification tasks
+"""
+from dgl.nn.pytorch import GatedGraphConv, GlobalAttentionPooling
+import torch
+from torch import nn
+import dgl
```
Comment (mufeili) [inline examples/pytorch/ggnn/ggnn_gc.py]: `dgl` hasn't been used.

File: examples/pytorch/ggnn/train_ns.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    test_acc_mean = np.mean(test_acc_list)
+    test_acc_std = np.std(test_acc_list)
+
+    print(f'Mean of accuracy in 10 test datasets: {test_acc_mean}, std: {test_acc_std}')
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Gate Graph Neural Networks for bAbI tasks')
+    parser.add_argument('--gpu', type=int, default=-1,
+                        help='gpu')
```
Comment (mufeili) [inline examples/pytorch/ggnn/train_ns.py]: - "Gate" should be "Gated"
- The `gpu` argument has not been used. Probably you can just remove it.

The issues above also occur in "train_gc.py" and "train_path_finding.py".

File: examples/pytorch/ggnn/train_path_finding.py [hunk 9]
```diff
@@ -91,21 +91,19 @@ def main(args):
 
 
 if __name__ == '__main__':
-    parser = argparse.ArgumentParser(description='Gate Graph Neural Networks for bAbI tasks')
-    parser.add_argument('--gpu', type=int, default=-1,
-                        help='gpu')
+    parser = argparse.ArgumentParser(description='Gated Graph Sequence Neural Networks for sequencial output tasks in '
```
Comment (mufeili) [inline examples/pytorch/ggnn/train_path_finding.py]: "sequencial" should be "sequential".

File: examples/pytorch/ggnn/README.md [hunk 10]
```diff
@@ -7,56 +7,178 @@
...
+The paper uses *Gated Graph Sequence Neural Networks (GGS-NNs)* to solve this kind of problems.
+In this example, we implemented GGS-NNs in `ggsnn.py`, run the following command for training 
+and testing:
+```bash
+python train_path_finding.py --train_num=250 --epochs=200
+```
 
+#### Results
 
+Following the paper, we use 10 different test sets for evaluation. The result is the the mean and
```
Comment (mufeili) [inline examples/pytorch/ggnn/README.md]: duplicate "the"

File: examples/pytorch/ggnn/README.md [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+form. In this example, we follow the paper to generate the datasets using symbolic form.
+There are 20 tasks in bAbI, in this example, we follow the paper to do task 4, 15, 16, 18 and 19.
+
+#### Task 4: Two argument relations: subject vs. object
+
+An example of task 4 is as follows
+```
+1 C e A
+2 A e B
+3 eval A w	C
```
Comment (mufeili) [inline examples/pytorch/ggnn/README.md]: Seems there's an issue for the space.

File: examples/pytorch/ggnn/README.md [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+
+A, B, C are nodes; e, w are edges, there are totally four kinds of edges: `n, s, w, e`, which can 
+be viewed as north, south, west, east.
+
+The first two lines are conditions, and the third line are the question and answer. 
+So the explanation of the example is:
+```
+1 Go east from C, we can reach A
+2 Go east from A, we can reach B
+3 Qustion: where can we reach if we go west from A? Answer: C
```
Comment (mufeili) [inline examples/pytorch/ggnn/README.md]: "Qustion" should be "Question"
